### ISSUE TITLE
coroutine generator use-after-free

### ISSUE THREAD

**Issue by xemul** [scylladb/seastar#1677]
Consider a code (can be added as a test case to coroutines_test.cc):

```c++
struct wrapped_string {
    sstring s;
};

coroutine::experimental::generator<wrapped_string, buffered_container> generate_wrapped_strings(coroutine::experimental::buffer_size_t size, int nr) {
    for (int i = 0; i < nr; i++) {
        sstring x = format("abcdefghijklmnopqrstuvwxyz-{}", i);
        co_yield wrapped_string(std::move(x));
    }
}

SEASTAR_TEST_CASE(test_generator_lref) {
    auto sgen = generate_wrapped_strings(coroutine::experimental::buffer_size_t{8}, 42);
    while (auto ws = co_await sgen()) {
        fmt::print("{}\n", ws->s);
    }
}
```

First, it's important that `sstring x` is _that_ long -- it must be external.

The bug -- if run in debug mode it crashes with heap use-after-free message:

* the memory was allocated where `sstring x` variable was declared -- that is sstring's external memory buffer.
* the memory was freed at the `generate_wrapped_string()`'s closing brace -- that is when the `x` went out of scope of the generator
* the use-after-free happens when `ws->s` is printed in the test case function

It seems that `ws->s` should be eventually copy-constructed or move-constructed from `x`, but in fact `ws` and `x` share the pointer to the external buffer for some reason.

There are two options how to work around that bug.

1. Declare constructor for `wrapped_string`

```c++
struct wrapped_string {
    sstring s;
    wrapped_string(sstring s_) : s(std::move(s_)) {}
};
```

2. Make `wrapped_string` local variable in generator

```c++
    for (int i = 0; i < nr; i++) {
        sstring x = format("abcdefghijklmnopqrstuvwxyz-{}", i);
        auto ws = wrapped_string(std::move(x));
        co_yield ws;
    }
```

cc @avikivity  @tchaikov

**Comment by avikivity** [comment]
Looks like a miscompile, no? with what clang do you see it?

**Comment by xemul** [comment]
> Looks like a miscompile, no? 

Yes. `ws` and `x` have different `this` values, but strings inside point to the same malloc-ed buffer. It all can be explained by "compiler memcpy-ed `x` into `ws` instead of copy/move constructing it"

> with what clang do you see it?

```
$ clang++ --version
clang version 15.0.7 (Fedora 15.0.7-2.fc37)
```

**Comment by avikivity** [comment]
Well I see a similar bug, but with clang 16 (and 15 is immune for my test case).

Try 16, maybe the fix that broke my test case fixes yours.

**Comment by tchaikov** [comment]
yeah, i am using clang-17, this test survived with the following steps:

```console
$ cmake -DCMAKE_CXX_COMPILER=clang++ -GNinja -Bbuild/debug -DCMAKE_BUILD_TYPE=Debug
$ ninja -C build/debug test_unit_coroutines
$ build/debug/tests/unit/coroutines_test
```

**Comment by bhalevy** [comment]
Can this issue be closed?
